text/2021-12-09-logging.md

```diff
@@ -0,0 +1,302 @@
+library users with a certain logging framework.
+
+Only use "high level" `V(n).Info() n > 0` in the library codes, do not use
+`Error()`, `V(0).Info()` and `Info()`. Because `V(0).Info()` (or not defined a
+level) and `Error()` would most probably print something to `stdout`, which
+would might break the design of cli tools.
+
+### Logger name
+
+The name of the logger should not be defined by itself, also should not be
+defined in the constructor. It should be defined in the caller for the
+constructor.
+
+GOOD Example:
+
+```go
+func main(){
+  val logger = initializeApplicationRootLogger()
+  NewWebModule(logger.WithName("web")).StartServe()
+}
+```
+
+BAD Example:
+
+```go
+func NewWebModule(logger logr.Logger) *WebModule {
+  return &WebModule{
+    logger: logger.WithName("web"),
+  }
+}
+```
+
+### Which level should I use
+
+TL;DR, resolving error with `Error()`; Within `Info()`, 0 is info and warning
+level, 1-4 is debug level, >=5 is trace level.
+
+We do not restrict the level of the logger, here is a list for the level we
+recommend to use:
+
+- Error() - Error should be used to indicated unexpected error, for example,
+  unexpected errors returned by subroutine function calls.
+
+- Info() - Info should be used as a human journal or diary. And it could be used
+  to log expected error as warning.
+
+- Info() hash multiple levels:
+  - V(0) - This is the default level, would ALWAYS be visible to users.
+    - CLI argument handling, print args before the program starts
+    - Application configuration
+    - Expected errors as warning message
+    - Journal for application's major behavior
+      - service endpoint: resolve requests
+      - long-time batch-job: job start, job complete
+    - Acquirement on system resources: listen on port, persistent in file, etc.
+  - V(1) - Default log level for debug.
+    - Expected error that repeat frequently, relates to conditions would be
+      corrected, like `StatusReasonConflict` when reconciling
+    - Component configuration and initialize
+  - V(2) - Useful state changes and conditional branch.
+    - State which is useful for debugging changed
+    - Choose a branch in a conditional statement
+  - V(3) - Extended information about changes
+    - Logging with state updates like `err := updateState(); err == nil`
+    - Logging error with more context before wrap and return it
+  - V(4) - Debug level verbosity, any other behavior that changes the state
+  - V(5) - Default log level for trace.
+    - Progressing in a for-loop
+    - Omitted things within "best effort" pattern
+    - Use default value with not configured value
+  - V(6) - Communication with other components
+    - Logging in handler before processing/resolving request
+    - RPC call with other components/services
+  - V(7) - Detailed information about communication
+    - Detailed payload and status of request/response within RPC
+
+The number of level still limited by the backend logging framework, for example,
```

Comment:
I think there are too many levels, which will make developers confused when using it in practice, could we reduce some levels here?

Reply:
I think we could use `V(0)` as default info level, `V(1)` as default debug level, and `V(5)` as default tracing level.

How do you think about it? @iguoyr

Reply:
If you agree with it, let's update this rfc, appending this introduction.

Reply:
Good idea!